### DISCUSSION TITLE
How do I create a formulae to install a bash shell script?

### DISCUSSION THREAD

**Discussion by etrigan63** [Homebrew/discussions#5388] [Writing Formulae/Casks]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.1-13-gb9a9b0a
ORIGIN: https://github.com/Homebrew/brew
HEAD: b9a9b0a5d10e3dda8148fde434456441c4db0f52
Last commit: 14 hours ago
Core tap HEAD: eb7d009e5dcfb6ad5abebe5a6865e89aae9f4138
Core tap last commit: 36 minutes ago
Core tap JSON: 22 May 13:33 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 16
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.1 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/bin/ruby
CPU: 16-core 64-bit zen3
Clang: N/A
Git: 2.45.1 => /bin/git
Curl: 8.6.0 => /bin/curl
Kernel: Linux 6.8.9-300.fc40.x86_64 x86_64 GNU/Linux
OS: Fedora Linux 40.20240518.0 (Bluefin)
Host glibc: 2.39
/usr/bin/gcc: 14.1.1
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have uncommitted modifications to Homebrew/homebrew-core.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core && git stash -u && git clean -d -f

Uncommitted files:
  ?? Formula/b/betterdiscordctl.rb
```

### Description of issue

Tried to write a formulae to install betterdiscordctl from https://github.com/bb010g/betterdiscordctl/archive/refs/tags/v2.0.1.tar.gz

Got the following errors running the audit:

HOMEBREW_NO_INSTALL_FROM_API=1 brew audit --new betterdiscordctl
Error: invalid attribute for formula 'betterdiscordctl': version (nil)
Warning: Removed Sorbet lines from backtrace!
Rerun with `--verbose` to see the original backtrace
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:330:in `validate_attributes!'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:255:in `initialize'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formulary.rb:570:in `new'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formulary.rb:570:in `get_formula'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formulary.rb:794:in `get_formula'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formulary.rb:1009:in `factory'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:146:in `block in load_formula_or_cask'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/api.rb:197:in `with_no_api_env'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/api.rb:206:in `with_no_api_env_if_needed'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:138:in `load_formula_or_cask'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:81:in `block in to_formulae_and_casks'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:79:in `each'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:79:in `flat_map'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:79:in `to_formulae_and_casks'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cli/named_args.rb:109:in `to_formulae_and_casks_with_taps'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/dev-cmd/audit.rb:156:in `block in run'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/api.rb:197:in `with_no_api_env'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/dev-cmd/audit.rb:126:in `run'
/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:92:in `<main>'
Please report this issue:
  https://docs.brew.sh/Troubleshooting

**Comment by p-linnane** [comment]
Difficult to troubleshoot without seeing the formula. Most likely you need `bin.install "betterdiscordctl"` in your `install` stanza.

**Reply by etrigan63** [reply]
```
class Betterdiscordctl < Formula
  desc "Utility for managing BetterDiscord on Linux"
  homepage ""
  url "https://github.com/bb010g/betterdiscordctl"
  sha256 "d56eaa43119bfeef96d842fb83ac943418389dd60d1a372a0e5e24e7eb68fecb"
  license "MIT"

  depends_on "cmake" => :build

  def install
    # Remove unrecognized options if they cause configure to fail
    # https://rubydoc.brew.sh/Formula.html#std_configure_args-instance_method
    system "./configure", "--disable-silent-rules", *std_configure_args
    # system "cmake", "-S", ".", "-B", "build", *std_cmake_args
  end

  test do
    system "false"
  end
end
```
For the record, this is the first time I have ever tried to create a brew formulae.

**Reply by SMillerDev** [reply]
Yeah, the `def install` should perform the steps needed to actually install the software. In your case I guess it's just `bin.install "file.sh" => "command_name"`.

**Reply by etrigan63** [reply]
OK, so this is what I have now and it passes the audit:
```
class Betterdiscordctl < Formula
  desc "Utility for managing BetterDiscord on Linux"
  homepage "https://github.com/bb010g/betterdiscordctl"
  url "https://github.com/bb010g/betterdiscordctl"
  version "2.0.1"
  sha256 "d56eaa43119bfeef96d842fb83ac943418389dd60d1a372a0e5e24e7eb68fecb"
  license "MIT"

  def install
    system "./configure", "--disable-silent-rules", *std_configure_args
    bin.install "https://github.com/bb010g/betterdiscordctl/archive/refs/tags/v2.0.1.tar.gz" => "betterdiscordctl"
  end

  test do
    system "false"
  end
end
```
What do I do now?

**Reply by p-linnane** [reply]
I've written a formula for you:
```
class Betterdiscordctl < Formula
  desc "Utility for managing BetterDiscord"
  homepage "https://github.com/bb010g/betterdiscordctl"
  url "https://github.com/bb010g/betterdiscordctl/archive/refs/tags/v2.0.1.tar.gz"
  sha256 "d56eaa43119bfeef96d842fb83ac943418389dd60d1a372a0e5e24e7eb68fecb"
  license "MIT"

  def install
    bin.install "betterdiscordctl"
  end

  test do
    assert_match version.to_s, shell_output(bin/"betterdiscordctl --version 2>&1")
  end
end
```

**Reply by etrigan63** [reply]
Awesome! Now how do I add it to Homebrew.sh so everyone else can enjoy it?

**Reply by p-linnane** [reply]
You don't, because it is does not meet [our criteria for inclusion](https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff). Please consider hosting this in [your own Homebrew tap](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap) instead.
